### ISSUE TITLE
Interest in moving to RNPM plugin??

### ISSUE THREAD

**Issue by jtag05** [ptmt/react-native-macos#186]
I was wondering if there is any interest in moving from a full blown fork of `react-native` to a plugin approach like `react-native-windows` adopted. My current project requirements make utilizing this library rather difficult as it stands. While I'm sure I can finagle a solution to my issues, wrestling with the fork approach has piqued my interest in maybe trying my hand at adopting this package into an rnpm plugin. Any feedback would be much appreciated.

**Comment by ericvicenti** [comment]
One thing to note, its rumored that Apple will soon make it much easier to run iOS apps on macOS. We might want to wait for the 2018 WWDC cycle to see what they introduce, because it might make our lives a lot easier.

**Comment by ptmt** [comment]
Yes, this is my intention. The future of this project has 3 possible ways, there is a project with implementing cross-platform React Native based on Qt, there is the Microsoft team working on new modular architecture for React Native, and there is Apple supposedly working on "rescuing the market of desktop macOS apps". We might finally have a superset of UIKit and AppKit or something like this.
Meantime I almost finished latest merging and it's been really exhausting, so it's not probably the right way how to do it. Even if Apple doesn't show anything, I'll spend some time on the new architecture.

**Comment by Kudo** [comment]
Updated rumor https://daringfireball.net/2018/04/scuttlebutt_regarding_ui_project.
Declarative UI ! hope it's not only human readable storyboard & xib. 

BTW, I was previously working on a new solution to port React Native into macOS.
As current react-native-macos's model is fork and modify code directly, and React Native upstream is actively development.
It would be a disaster to follow upstream.

I was proposing a layer called RCTXUIKit which is somewhat [Chameleon](http://chameleonproject.org/) did.
I tried to use ObjC category or inheritance to make AppKit's interface compatible with UIKit.
From React Native perspective,

1. Only to import RCTXUIKit.h for most of code.
2. Adding  macOS support into React.xcodeproj, pretty like tvOS did.

This model should be maintainable with upstream, more than that I also hope if it could be merged into upstream.
My current status is in early stage which still in fixing build break.
I have no experience for macOS development still not sure if this model will works.
Do you have any thought for this proposal?

**Comment by keplersj** [comment]
You all might be interested in this post-WWDC experiment done with React Native and UIKit for Mac (Marzipan): [notjosh/Marzipants](https://github.com/notjosh/Marzipants), and this GitHub issue: facebook/react-native#19559.

**Comment by Kudo** [comment]
Awesome @keplersj !
It's so surprising that the POC works without patching much code.
